flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -0,0 +1,216 @@
+            SimpleVersionedSerializer<? extends T> s) {
+        @SuppressWarnings("rawtypes")
+        SimpleVersionedSerializer s1 = s;
+        return s1;
+    }
+
+    /** Serializes splits by delegating to the source-indexed split serializer. */
+    public static class SplitSerializerWrapper<SplitT extends SourceSplit>
+            implements SimpleVersionedSerializer<HybridSourceSplit> {
+
+        final List<SimpleVersionedSerializer<SourceSplit>> serializers;
+
+        public SplitSerializerWrapper(List<SimpleVersionedSerializer<SourceSplit>> serializers) {
+            this.serializers = serializers;
+        }
+
+        @Override
+        public int getVersion() {
+            return 0;
+        }
+
+        @Override
+        public byte[] serialize(HybridSourceSplit split) throws IOException {
+            try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    DataOutputStream out = new DataOutputStream(baos)) {
+                out.writeInt(split.sourceIndex());
+                out.writeInt(serializerOf(split.sourceIndex()).getVersion());
+                byte[] serializedSplit =
+                        serializerOf(split.sourceIndex()).serialize(split.getWrappedSplit());
+                out.writeInt(serializedSplit.length);
+                out.write(serializedSplit);
+                return baos.toByteArray();
+            }
+        }
+
+        @Override
+        public HybridSourceSplit deserialize(int version, byte[] serialized) throws IOException {
+            try (ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
+                    DataInputStream in = new DataInputStream(bais)) {
+                int sourceIndex = in.readInt();
+                int nestedVersion = in.readInt();
+                int length = in.readInt();
+                byte[] splitBytes = new byte[length];
+                in.readFully(splitBytes);
+                SourceSplit split =
+                        serializerOf(sourceIndex).deserialize(nestedVersion, splitBytes);
+                return new HybridSourceSplit(sourceIndex, split);
+            }
+        }
+
+        private SimpleVersionedSerializer<SourceSplit> serializerOf(int sourceIndex) {
+            Preconditions.checkArgument(sourceIndex < serializers.size());
+            return serializers.get(sourceIndex);
+        }
+    }
+
+    /**
+     * Converts checkpoint between sources to transfer end position to next source's start position.
+     * Only required for dynamic position transfer at time of switching, otherwise source can be
+     * preconfigured with a start position during job submission.
+     */
+    public interface CheckpointConverter<InCheckpointT, OutCheckpointT>
+            extends Function<InCheckpointT, OutCheckpointT>, Serializable {}
+
+    /** Chain of sources with option to convert start position at switch-time. */
+    public static class SourceChain<T, SplitT extends SourceSplit, EnumChkT>
```

Comment:
Related to earlier comment on repurposing `SourceChain` to `Builder`, I found the generic types are a little daunting. Maybe the user-facing builder just exposes the `add(...)` API. Can we avoid the generic types for builder class this way and keep those type complexity internal?

Reply:
The type parameters are required to provide compile time checking so that only source with compatible record type and state type can be combined. The state type should also match of the user provided a converter function.

I removed `SplitT`, it wasn't needed.